c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -3702,6 +3704,382 @@ bool init_class(JSContext *cx, HandleObject global) {
+    return JS::Result<std::string>(JS::Error());
+  case 0:
+    break;
+  default:
+    MOZ_CRASH("Too many characters leftover");
+  }
+
+  return output;
+}
+
+// https://html.spec.whatwg.org/multipage/webappapis.html#dom-atob
+bool atob(JSContext *cx, unsigned argc, Value *vp) {
+  CallArgs args = CallArgsFromVp(argc, vp);
+  if (!args.requireAtLeast(cx, "atob", 1)) {
+    return false;
+  }
+  auto dataResult = ConvertJSValueToByteString(cx, args.get(0));
+  if (dataResult.isErr()) {
+    return false;
+  }
+  auto data = dataResult.unwrap();
+
+  // 1. Let decodedData be the result of running forgiving-base64 decode on data.
+  auto decoded_result = forgivingBase64Decode(data);
+  // 2. If decodedData is failure, then throw an "InvalidCharacterError" DOMException.
+  if (decoded_result.isErr()) {
+    JS_ReportErrorNumberUTF8(cx, GetErrorMessage, nullptr, JSMSG_INVALID_CHARACTER_ERROR);
+    return false;
+  }
+  auto decoded = decoded_result.unwrap();
+
+  RootedString decodedData(cx, JS_NewStringCopyN(cx, decoded.c_str(), decoded.length()));
+  if (!decodedData) {
+    return false;
+  }
+
+  // 3. Return decodedData.
+  args.rval().setString(decodedData);
+  return true;
+}
+
+const char base[65] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+                      "abcdefghijklmnopqrstuvwxyz"
+                      "0123456789+/";
+
+inline uint8_t CharTo8Bit(char character) { return uint8_t(character); }
+inline std::string base64Encode3to4(std::string aSrc) {
+  std::string out = "";
+  uint32_t b32 = 0;
+  int i, j = 18;
+
+  for (i = 0; i < 3; ++i) {
+    b32 <<= 8;
+    b32 |= CharTo8Bit(aSrc[i]);
+  }
+
+  for (i = 0; i < 4; ++i) {
+    out += base[(uint32_t)((b32 >> j) & 0x3F)];
+    j -= 6;
+  }
+  return out;
+}
+
+inline std::string base64Encode2to4(std::string aSrc) {
+  std::string out = "";
+  uint8_t src0 = CharTo8Bit(aSrc[0]);
+  uint8_t src1 = CharTo8Bit(aSrc[1]);
+  out += base[(uint32_t)((src0 >> 2) & 0x3F)];
+  out += base[(uint32_t)(((src0 & 0x03) << 4) | ((src1 >> 4) & 0x0F))];
+  out += base[(uint32_t)((src1 & 0x0F) << 2)];
+  out += '=';
+  return out;
+}
+
+inline std::string base64Encode1to4(std::string aSrc) {
+  std::string out = "";
+  uint8_t src0 = CharTo8Bit(aSrc[0]);
+  out += base[(uint32_t)((src0 >> 2) & 0x3F)];
+  out += base[(uint32_t)((src0 & 0x03) << 4)];
+  out += '=';
+  out += '=';
+  return out;
+}
+
+// https://infra.spec.whatwg.org/#forgiving-base64-encode
+// To forgiving-base64 encode given a byte sequence data, apply the base64 algorithm defined in
+// section 4 of RFC 4648 to data and return the result. [RFC4648] Note: This is named
+// forgiving-base64 encode for symmetry with forgiving-base64 decode, which is different from the
+// RFC as it defines error handling for certain inputs.
+std::string forgivingBase64Encode(std::string data) {
+  int length = data.length();
+  std::string out = "";
+  while (length >= 3) {
+    out += base64Encode3to4(data);
+    data.erase(0, 3);
+    length -= 3;
+  }
```

Comment:
Again, I'd refactor this to use a `std::string_view` so that you're not copying the buffer around while encoding it, and not paying any cost for moving the pointer at the front of `input` around.